### PR TITLE
simple blackduck plugin to validate high violations in a build

### PR DESCRIPTION
**Pull request by rmannibucau**
Simple way to prevent a release if too much violations (HIGH for release and components ATM) are encountered

### CODE REVIEW

File: talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckBase.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+        if (blackduckUrl == null) {
+            getLog().error("No url specified, please set blackduckUrl");
+            return;
+        }
+
+        final Optional<Server> serverOpt = session.getSettings().getServers().stream().filter(s -> serverId.equals(s.getId()))
+                .findFirst();
+        if (!serverOpt.isPresent()) {
+            getLog().warn("No server '" + serverId + "', skipping blackduck execution");
```
Comment (talend-glorieux) [inline talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckBase.java]: Please use a string formater.

File: talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckBase.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    protected List<MavenProject> reactorProjects;
+
+    @Component
+    protected SettingsDecrypter settingsDecrypter;
+
+    @Override
+    public void execute() throws MojoExecutionException, MojoFailureException {
+        final AtomicInteger counter = AtomicInteger.class.cast(
+                session.getRequest().getData().computeIfAbsent(getClass().getName() + ".counter", k -> new AtomicInteger()));
+        if (atTheEnd && !skip /* if skipped log the message */ && counter.incrementAndGet() != reactorProjects.size()) {
```
Comment (talend-glorieux) [inline talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckBase.java]: Please avoid adding comments in the middle of code.

File: talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckBase.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            getLog().debug("Not yet at the last project, will only run when reached to not do it multiple times " + counter.get()
```
Comment (talend-glorieux) [inline talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckBase.java]: Please use a string formater.

File: talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckBase.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        final Optional<Server> serverOpt = session.getSettings().getServers().stream().filter(s -> serverId.equals(s.getId()))
+                .findFirst();
+        if (!serverOpt.isPresent()) {
+            getLog().warn("No server '" + serverId + "', skipping blackduck execution");
+            return;
+        }
+
+        Server server = serverOpt.get();
+        if ("skip".equals(server.getPassword())) {
+            getLog().warn("server '" + serverId + "' was configured to be skipped");
```
Comment (talend-glorieux) [inline talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckBase.java]: Please use a string formater.

File: talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckValidateMojo.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+                    new ReportCategoriesEnum[] { ReportCategoriesEnum.COMPONENTS });
+        } catch (IntegrationException e) {
+            throw new MojoExecutionException(e.getMessage(), e);
+        }
+
+        {
+            final int operationalHigh = report.getAggregateBomViewEntries().stream()
+                    .mapToInt(it -> it.getRiskProfile().getCategories().getOPERATIONAL().getHIGH()).sum();
+            if (operationalHigh >= acceptedOperationalHigh) {
+                final String message = "Found #" + operationalHigh + " operational high violations, accepted: #"
```
Comment (talend-glorieux) [inline talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckValidateMojo.java]: Please use a string formater.

File: talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckValidateMojo.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+                        + acceptedOperationalHigh;
+                getLog().error(message);
+                throw new MojoFailureException(message);
+            }
+        }
+        {
+            final int licenseHigh = report.getAggregateBomViewEntries().stream().mapToInt(it -> it.getLicenseRisk().getHIGH())
+                    .sum();
+            if (licenseHigh >= acceptedLicenseRiskHigh) {
+                final String message = "Found #" + licenseHigh + " license violations, accepted: #" + acceptedLicenseRiskHigh;
```
Comment (talend-glorieux) [inline talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckValidateMojo.java]: Please use a string formater.

File: talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/HubDetectMojo.java [hunk 7]
```diff
@@ -264,7 +180,9 @@ public void execute() throws MojoExecutionException, MojoFailureException {
                         + "\",\n" + "\"blackduck.hub.password\": \"" + server.getPassword() + "\",\n"
                         + "\"logging.level.com.blackducksoftware.integration\": \"" + logLevel + "\",\n"
                         + "\"detect.project.name\": \"" + blackduckName + "\",\n" + "\"detect.source.path\": \""
-                        + root.getAbsolutePath() + "\"\n}");
+                        + rootProject.getBasedir().getAbsolutePath() + "\"\n}");
```
Comment (talend-glorieux) [inline talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/HubDetectMojo.java]: Please don't build JSONs by hand.

File: talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckValidateMojo.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        final ProjectDataService projectService = hsf.createProjectDataService();
+        ProjectVersionWrapper project;
+        try {
+            project = projectService.getProjectVersion(blackduckName, mvnProject.getVersion());
+        } catch (final IntegrationException e) {
+            if (!mvnProject.getVersion().endsWith("-SNAPSHOT")) { // use the closest version, ie the snapshot
+                try {
+                    project = projectService.getProjectVersion(blackduckName, mvnProject.getVersion() + "-SNAPSHOT");
+                } catch (final IntegrationException e1) {
+                    try {
```
Comment (fhuaulme) [inline talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckValidateMojo.java]: You don't log `e1` nor `e2` (or rethrow it), is it on purpose?

File: talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckValidateMojo.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+        }
+        final VersionReport report;
+        try {
+            report = reportService.generateHubReport(project.getProjectVersionView(), ReportFormatEnum.JSON,
+                    new ReportCategoriesEnum[] { ReportCategoriesEnum.COMPONENTS });
+        } catch (IntegrationException e) {
+            throw new MojoExecutionException(e.getMessage(), e);
+        }
+
+        {
```
Comment (fhuaulme) [inline talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckValidateMojo.java]: The next 3 code blocks are somewhat duplicated blocks, what changes it - as far as I can tell -:
* The mapToInt function
* The message

File: talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckValidateMojo.java (hunk 8 above)
Comment (rmannibucau) [reply talend-tools-maven-plugin/src/main/java/org/talend/tools/blackduck/BlackduckValidateMojo.java]: yep, don't care much and throw the root case in this case

**Pull request merged**